src/dataflow/src/render/debezium.rs

```diff
@@ -0,0 +1,661 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::cmp::max;
+use std::collections::{HashMap, HashSet};
+use std::str::FromStr;
+
+use chrono::format::{DelayedFormat, StrftimeItems};
+use chrono::NaiveDateTime;
+use differential_dataflow::AsCollection;
+use differential_dataflow::Collection;
+use log::{debug, error, info, warn};
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::Operator;
+use timely::dataflow::{Scope, Stream};
+
+use dataflow_types::{
+    sources::{DebeziumDedupProjection, DebeziumEnvelope, DebeziumMode, DebeziumSourceProjection},
+    DataflowError,
+};
+use expr::GlobalId;
+use repr::{Datum, Diff, Row};
+
+use crate::metrics::Metrics;
+use crate::source::DecodeResult;
+
+pub(crate) fn render<G: Scope>(
+    envelope: &DebeziumEnvelope,
+    input: &Stream<G, DecodeResult>,
+    debug_name: String,
+    metrics: Metrics,
+    src_id: GlobalId,
+    dataflow_id: usize,
+) -> Collection<G, Result<Row, DataflowError>, Diff> {
+    let (before_idx, after_idx) = (envelope.before_idx, envelope.after_idx);
+    match envelope.mode {
+        DebeziumMode::Upsert => {
+            let gauge = metrics.debezium_upsert_count_for(src_id, dataflow_id);
+            input
+                .unary(Pipeline, "envelope-debezium-upsert", move |_, _| {
+                    let mut current_values = HashMap::new();
+                    let mut data = vec![];
+                    move |input, output| {
+                        gauge.set(current_values.len() as u64);
```

Comment:
This is just tracking the number of keys in the upsert state and as far as I understand it is doing the same thing as before https://github.com/MaterializeInc/materialize/blob/main/src/dataflow/src/decode/mod.rs#L61-L62

Are you maybe asking why the gauge is set in the entrypoint of the operator closure instead of the loop of the input? (If yes then I agree that I should move it there)

Reply:
I was more asking about the first thing, but I now also agree it should be moved